### PR TITLE
update session token password on user password change

### LINKED ISSUE #24544: Changing own password logs you out
#24189 creates and validates browser sessions depending on an entry in the database where the existence of that token is checked once a minute and the validity of the password is checked once every five minutes. While it is expected that clients and other browser sessions are disconnected when changing the user password, this will also cause a *forced logout* of the browser session changing the password.
This should be prevented by updating the database entry with the new password.

cc @LukasReschke

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes https://github.com/owncloud/core/issues/24544

@PVince81 @DeepDiver1975 @rullzer @nickvergessen review please

### CODE REVIEW

File: lib/private/User/Session.php [hunk 1]
```diff
@@ -676,4 +676,21 @@ public function unsetMagicInCookie() {
 		setcookie('oc_remember_login', '', time() - 3600, OC::$WEBROOT . '/', '', $secureCookie, true);
 	}
 
+	/**
+	 * Update password of the browser session token if there is one
+	 *
+	 * @param string $password
+	 */
+	public function updateSessionTokenPassword($password) {
```
Comment (PVince81) [inline lib/private/User/Session.php]: as discussed, not added to the interface for now

File: lib/private/Authentication/Token/IProvider.php [hunk 2]
```diff
@@ -99,4 +99,14 @@ public function getTokenByUser(IUser $user);
...
+
+	/**
+	 * Encrypt and set the password of the given token
+	 *
+	 * @param IToken $token
+	 * @param string $tokenId
+	 * @param string $password
+	 * @throws InvalidTokenException
+	 */
+	public function setPassword(IToken $token, $tokenId, $password);
```
Comment (PVince81) [inline lib/private/Authentication/Token/IProvider.php]: in general we don't change interfaces after feature freeze, but I don't think people started writing token providers in such a short time, so this is ok

Comment (ChristophWurst) [inline lib/private/Authentication/Token/IProvider.php]: and this is a private interface. Nobody should use it except core.

Comment (ChristophWurst) [inline lib/private/Authentication/Token/IProvider.php]: additionally we do not (yet) allow to plug custom token providers into core

**Comment by PVince81** [comment]
Tested, doesn't work.

I changed a user's password while on this branch then waited around 5 minutes, then clicked the ownCloud logo to navigate to the main page => this returns to the login page

**Comment by PVince81** [comment]
My steps:

1. Create a user "user1"
1. Login as "user1"
1. Go to personal page
1. Change the password of "user1"
1. Wait 5 minutes
1. Click ownCloud logo

Will retry this again

**Comment by ChristophWurst** [comment]
confirmed, for some reason it stopped working.

**Comment by ChristophWurst** [comment]
oh, I updated the password with the username. When testing with username===password it worked :see_no_evil: 

@PVince81 please re-review

**Comment by PVince81** [comment]
Tested, works now :+1:

**Comment by icewind1991** [comment]
Looks good :+1:

**Pull request merged**